Add policy verdicts GSG (#12165)
cilium-team/cilium: Documentation/gettingstarted/policy-creation.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _policy_verdicts:

*******************************
Creating policies from verdicts
*******************************

Policy Audit Mode configures Cilium to allow all traffic while logging all

> connections that would otherwise be dropped by policy. Policy audit mode may be

Reviewer: just to be consistent:
```suggestion
connections that would otherwise be dropped by policy. Policy Audit Mode may be
```

          k8s:io.cilium.k8s.policy.serviceaccount=default
          k8s:io.kubernetes.pod.namespace=default
          k8s:org=empire

Given the above information, we now know the labels of the target pod, the
labels of the peer that's attempting to connect, the direction of the traffic
and the port. In this case, we can see clearly that it's an empire craft
so once we've determined that we expect this traffic to arrive at the
deathstar, we can form a policy to match the traffic:

.. literalinclude:: ../../examples/minikube/sw_l3_l4_policy.yaml

To apply this L3/L4 policy, run:

.. parsed-literal::

    $ kubectl create -f \ |SCM_WEB|\/examples/minikube/sw_l3_l4_policy.yaml
    ciliumnetworkpolicy.cilium.io/rule1 created

Now if we run the landing requests again, we can observe in the monitor output
that the traffic which was previously audited to be dropped by the policy are
now reported as allowed:

.. code-block:: shell-session

    $ kubectl exec tiefighter -- curl -s -XPOST deathstar.default.svc.cluster.local/v1/request-landing
    Ship landed

Executed from the cilium pod:

.. code-block:: shell-session

   # cilium monitor -t policy-verdict

>    Policy verdict log: flow 0xabf3bda6 local EP ID 232, remote ID 31028, dst port 80, proto 6, ingress true, action allow, match L3-L4, 10.0.0.112 :59824 -> 10.0.0.147:80 tcp SYN

Reviewer: would be nice to include the output from xwing as well:

```
# from tiefigher
Policy verdict log: flow 0x61744c95 local EP ID 484, remote ID 48203, dst port 80, proto 6, ingress true, action allow, match L3-L4, 10.0.0.107:38786 -> 10.0.0.113:80 tcp SYN

# from xwing
Policy verdict log: flow 0x2a950ee6 local EP ID 484, remote ID 44930, dst port 80, proto 6, ingress true, action audit, match none, 10.0.0.252:39122 -> 10.0.0.113:80 tcp SYN

Author: I thought about it, but the content is the same as from the tiefighter at the beginning. I'm not sure what this adds (other than using an extra pod that we deployed).

Author: Ah! How about this: We disable policy audit mode and validate that the policy works as intended. Coming soon:tm:

Author: ^^^ Done

Now the policy verdict states that the traffic would be allowed: ``action

> allow``. Success!

Author: I'd like to end this up a bit less abruptly but I'm running out of creative juice at the moment :-)

Reviewer: that's what you get for protecting the empire man.

so once we've determined that we expect this traffic to arrive at the
deathstar, we can form a policy to match the traffic:

.. literalinclude:: ../../examples/minikube/sw_l3_l4_policy.yaml

To apply this L3/L4 policy, run:

.. parsed-literal::

    $ kubectl create -f \ |SCM_WEB|\/examples/minikube/sw_l3_l4_policy.yaml
    ciliumnetworkpolicy.cilium.io/rule1 created

Now if we run the landing requests again, we can observe in the monitor output
that the traffic which was previously audited to be dropped by the policy are
now reported as allowed:

.. code-block:: shell-session

    $ kubectl exec tiefighter -- curl -s -XPOST deathstar.default.svc.cluster.local/v1/request-landing
    Ship landed

Executed from the cilium pod:

.. code-block:: shell-session

   # cilium monitor -t policy-verdict
   Policy verdict log: flow 0xabf3bda6 local EP ID 232, remote ID 31028, dst port 80, proto 6, ingress true, action allow, match L3-L4, 10.0.0.112 :59824 -> 10.0.0.147:80 tcp SYN

Now the policy verdict states that the traffic would be allowed: ``action
allow``. Success!

Disable Policy Audit Mode
=========================

> These steps should be repeated for each connection in the cluster to ensure

Reviewer: ```suggestion
These steps should be repeated for each node in the cluster to ensure
```

Reviewer: never mind. i think i misread this sentence.

            daemon set "cilium" successfully rolled out

      .. group-tab:: Helm Upgrade

         .. parsed-literal::

            helm upgrade cilium |CHART_RELEASE| \\
               --namespace $CILIUM_NAMESPACE \\
               --reuse-values \\
               --set config.policyAuditMode=false

Now if we run the landing requests again, only the *tiefighter* pods with the
label ``org=empire`` will succeed. The *xwing* pods will be blocked!

.. code-block:: shell-session

    $ kubectl exec tiefighter -- curl -s -XPOST deathstar.default.svc.cluster.local/v1/request-landing
    Ship landed

This works as expected. Now the same request run from an *xwing* pod will fail:

.. code-block:: shell-session

    $ kubectl exec xwing -- curl -s -XPOST deathstar.default.svc.cluster.local/v1/request-landing

This request will hang, so press Control-C to kill the curl request, or wait
for it to time out.

We hope you enjoyed the tutorial.  Feel free to play more with the setup,
follow the `gs_http` guide, and reach out to us on the `Cilium
Slack channel <https://cilium.herokuapp.com>`_ with any questions!

Clean-up
========

.. parsed-literal::

>    $ kubectl delete -f \ |SCM_WEB|\/examples/minikube/http-sw-app.yaml

Reviewer: this should work without backslashes: https://github.com/cilium/cilium/blame/1cc79c1fb522043ec8dd899285f9c0ba5cae8e88/Documentation/gettingstarted/quick-install.rst#L23

```suggestion
   $ kubectl delete -f |SCM_WEB|/examples/minikube/http-sw-app.yaml
   $ kubectl delete -f |SCM_WEB|/examples/minikube/sw_deny_policy.yaml
```

Author: I just copied & pasted & modified from other examples. Seems like 90% of the docs use this approach. We can follow up to make them all consistent and maybe drop the `\` if it's really not necessary.

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _policy_verdicts:

*******************************
Creating policies from verdicts
*******************************

> Policy Audit mode allows the user allow all traffic to pass through Cilium

Reviewer: maybe something like:
```suggestion
Cilium in Policy Audit mode allows all traffic to pass through while logging all connections that would otherwise be dropped by policy using policy verdict notifications.
```

while logging all connections that would otherwise be dropped by policy using
policy verdict notifications. Policy audit mode may be configured for the
entire daemon using ``--policy-audit-mode=true``. Layer 3 and 4 policies are
currently supported. When policy audit mode is enabled and you've installed a
policy that doesn't match traffic, ``cilium monitor`` prints the following:

.. code-block:: shell-session

   # cilium monitor -t policy-verdict
   ...

>    Policy verdict log: flow 0x63113709 local EP ID 1121, remote ID 2986, dst port 80, proto 6, ingress true, action audit, match none, 10.29.210.187:54134 -> 10.29.50.40:80 tcp SYN

Reviewer: not directly related to the documentation, but kind of curious why policy-verdict shows endpoint id for local and security identity for remote.

Author: Great question. @ap4y do you recall why this is?

Author: Actually this may just be historic in that all of the BPF notification types followed this approach.

Reviewer 2: I'm not entirely sure, @lzang probably knows more about the design decision for this one. AFAIK security identity is used during the [verdict resolution](https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1119-L1137) so it makes sense to expose it in the policy log. Source is set automatically based on the datapath via a common header of the message payload: https://github.com/cilium/cilium/blob/0fa8ac7f4310206e8ce7ecaf5276d785d5c0464b/bpf/lib/common.h#L270-L274 And we have source defined in the datapath implementation: https://github.com/cilium/cilium/blob/0fa8ac7f4310206e8ce7ecaf5276d785d5c0464b/bpf/bpf_lxc.c#L12

Author: Maybe makes sense for us to take another look at the CLI output and see whether there are some basic quality-of-life things like this to also print the ~source~ local identity.

>    Policy verdict log: flow 0x63113709 local EP ID 1121, remote ID 2986, dst port 80, proto 6, ingress true, action audit, match none, 10.29.210.187:54134 -> 10.29.50.40:80 tcp SYN

Reviewer 3: Once #12137 is merged, this example output should probably be updated to the new format.

> In the above example, we can see that traffic to endpoint ``11211`` has

Reviewer: ```suggestion
In the above example, we can see that endpoint ``11211`` has
```

> received traffic (``ingress true``) which doesn't match the policy (``action

Reviewer: is there a corresponding egress policy verdict notification in the node where remote is running?

Author: Yes:

```
Policy verdict log: flow 0x1c39ef18 local EP ID 969, remote ID 22111, dst port 80, proto 6, ingress false, action audit, match none, 10.29.112.245:60218 -> 10.29.179.216:80 tcp SYN
Policy verdict log: flow 0x1c39ef18 local EP ID 2239, remote ID 17138, dst port 80, proto 6, ingress true, action allow, match L3-L4, 10.29.112.245:60218 -> 10.29.179.216:80 tcp SYN
```

One trick I found while reviewing your feedback is that there needs to be a policy already for audit mode to take effect. If policy is not enabled, there won't be policy notifications AFAICT.

> audit match none``). The source of this traffic has the identity ``2986``.

Reviewer: does `match none` show up when there is no policy at all?

Author: Actually no, seems like policy verdicts don't show up if there is no policy at all.

                                                              k8s:io.cilium.k8s.policy.serviceaccount=default
                                                              k8s:io.kubernetes.pod.namespace=default
                                                              k8s:org=empire
   ...
   # cilium identity get 2986
   ID     LABELS
   2986   k8s:class=tiefighter
          k8s:io.cilium.k8s.policy.cluster=default
          k8s:io.cilium.k8s.policy.serviceaccount=default
          k8s:io.kubernetes.pod.namespace=default
          k8s:org=empire

Given the above information, we now know the labels of the target pod, the
labels of the peer that's attempting to connect, the direction of the traffic
and the port. We can now form a policy to match the traffic:

.. code-block:: shell-session

   $ cat sw_l3_l4_policy.yaml
   apiVersion: "cilium.io/v2"
   kind: CiliumNetworkPolicy
   description: "L3-L4 policy to restrict deathstar access to empire ships only"
   metadata:
     name: "rule1"
   spec:
     endpointSelector:
       matchLabels:
         org: empire
         class: deathstar
     ingress:
     - fromEndpoints:
       - matchLabels:

>            org: empire

Reviewer: maybe list `class=tiefighter` here to be consistent.

       toPorts:
       - ports:
         - port: "80"
           protocol: TCP
   $ kubectl create -f sw_l3_l4_policy.yaml
   ciliumnetworkpolicy.cilium.io/rule1 created

>    $ kubectl -n kube-system exec -ti $(get_cilium_pod) cilium monitor -t policy-verdict

Reviewer: `$(get_cilium_pod)` doesn't seem to be defined.

>    Policy verdict log: flow 0xabf3bda6 local EP ID 343, remote ID 2986, dst port 80, proto 6, ingress true, action allow, match L3-L4, 10.29.210.187:59824 -> 10.29.47.87:80 tcp SYN

Reviewer 3: Same here, this needs to be updated once #12137 is merged.
